core/Session/SessionAuth.php

```diff
@@ -0,0 +1,139 @@
+ */
+
+namespace Piwik\Session;
+
+use Piwik\Auth;
+use Piwik\AuthResult;
+use Piwik\Date;
+use Piwik\Plugins\UsersManager\Model as UsersModel;
+use Piwik\Session;
+
+/**
+ * Validates already authenticated sessions.
+ *
+ * See {@link \Piwik\Session\SessionFingerprint} for more info.
+ */
+class SessionAuth implements Auth
+{
+    /**
+     * For tests, since there's no actual session there.
+     *
+     * @var bool
+     */
+    private $shouldDestroySession;
+
+    public function __construct($shouldDestroySession = true)
+    {
+        $this->shouldDestroySession = $shouldDestroySession;
+    }
+
+    public function getName()
+    {
+        // empty
+    }
+
+    public function setTokenAuth($token_auth)
+    {
+        // empty
+    }
+
+    public function getLogin()
+    {
+        // empty
+    }
+
+    public function getTokenAuthSecret()
+    {
+        // empty
+    }
+
+    public function setLogin($login)
+    {
+        // empty
+    }
+
+    public function setPassword($password)
+    {
+        // empty
+    }
+
+    public function setPasswordHash($passwordHash)
+    {
+        // empty
+    }
+
+    public function authenticate()
+    {
+        $sessionId = new SessionFingerprint();
+        $userModel = new UsersModel();
+
+        $userForSession = $sessionId->getUser();
+        if (empty($userForSession)) {
+            return $this->makeAuthFailure();
+        }
+
+        $user = $userModel->getUser($userForSession);
+        if (empty($user)) {
+            return $this->makeAuthFailure();
+        }
+
+        if (!$sessionId->isMatchingCurrentRequest()) {
+            // this user should be using a different session, so generate a new ID
+            // NOTE: Zend_Session cannot be used since it will destroy the old
+            // session.
+            if ($this->shouldDestroySession) {
+                session_regenerate_id();
+            }
+            $sessionId->clear();
+
+            return $this->makeAuthFailure();
+        }
+
+        $tsPasswordModified = $user['ts_password_modified'];
+        if ($this->isSessionStartedBeforePasswordChange($sessionId, $tsPasswordModified)) {
+            // Note: can't use Session::destroy() since Zend prohibits starting a new session
+            // after session_destroy() is called.
+            $sessionId->clear();
+            if ($this->shouldDestroySession) {
+                Session::regenerateId();
+            }
+
+            return $this->makeAuthFailure();
+        }
+
+        return $this->makeAuthSuccess($user);
+    }
+
+    private function isSessionStartedBeforePasswordChange(SessionFingerprint $sessionId, $tsPasswordModified)
+    {
+        // if the session start time doesn't exist for some reason, log the user out
+        $sessionStartTime = $sessionId->getSessionStartTime();
+        if (empty($sessionStartTime)) {
+            return true;
+        }
+
+        return $sessionStartTime < Date::factory($tsPasswordModified)->getTimestampUTC();
```

Comment:
FYI: I currently got this error because `$tsPasswordModified ` is null. In my case it did not execute the DB update because of something else and now I cannot access the UI anymore. May be good to ignore this check if forever reason that password modified is not set.

![image](https://user-images.githubusercontent.com/273120/32355590-629086b6-c094-11e7-90dd-b9a0130044ce.png)

Reply:
The update probably has an old version, need to change it before it gets merged.